### PR TITLE
Allow external ownership of CHIPDeviceController operational key

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem
- For controllers that make use of their own key management,
  such as OS-aided key management, it is currently impossible
  to pass-in the Operational Key pair to use for a given controller.

Issue #18444
Issue #7695

#### Change overview
This PR adds APIs to avoid FabricTable from trying to manage
the lifecycle of the operational key, and allow the caller/initializer
of CHIPDeviceControllerFactory to properly inject its own operational
key that it manages, without attempts of storage or dynamic memory
management.

#### Testing
- Unit tests still pass, cert tests as well
- This was tested internally by us using the APIs to
  manage a key for Android. That change will be a follow-up

### CODE REVIEW

File: src/controller/CHIPDeviceController.h [hunk 1]
```diff
@@ -240,6 +247,13 @@ class DLL_EXPORT DeviceController : public SessionRecoveryDelegate, public Abstr
      */
     CHIP_ERROR DisconnectDevice(NodeId nodeId);
 
+    /**
+     * @brief
+     *   Configures a new set of operational credentials to be used with this
+     *   controller.
+     */
+    CHIP_ERROR InitControllerNOCChain(const ControllerInitParams & params);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.h:255]: Why is this being moved to a public API?  It needs to be called from `Init`, no?

File: src/credentials/FabricTable.cpp [hunk 2]
```diff
@@ -95,7 +95,7 @@ CHIP_ERROR FabricInfo::CommitToStorage(PersistentStorageDelegate * storage)
 
     {
         Crypto::P256SerializedKeypair serializedOpKey;
-        if (mOperationalKey != nullptr)
+        if (!mHasExternallyOwnedOperationalKey && mOperationalKey != nullptr)
```
Comment (bzbarsky-apple) [inline src/credentials/FabricTable.cpp:98]: But if `mHasExternallyOwnedOperationalKey` why are we wasting time creating a random keypair and setting up `serializedOpKey` if it will be ignored?  Seems like we should have something more like:
```suggestion
        if (!mHasExternallyOwnedOperationalKey) {
            if (mOperationalKey != nullptr) {
```
```
...
            } 
            else 
            {
                // random keypair bits here
            }
        }
```

File: src/credentials/FabricTable.cpp [hunk 3]
```diff
@@ -318,9 +321,12 @@ CHIP_ERROR FabricInfo::DeleteFromStorage(PersistentStorageDelegate * storage, Fa
     return prevDeleteErr;
 }
 
-CHIP_ERROR FabricInfo::SetOperationalKeypair(const P256Keypair * keyPair)
+CHIP_ERROR FabricInfo::SetOperationalKeypair(P256Keypair * keyPair)
```
Comment (bzbarsky-apple) [inline src/credentials/FabricTable.cpp:324]: Why this change?

File: src/credentials/FabricTable.cpp [hunk 4]
```diff
@@ -335,6 +341,20 @@ CHIP_ERROR FabricInfo::SetOperationalKeypair(const P256Keypair * keyPair)
     return mOperationalKey->Deserialize(serialized);
 }
 
+CHIP_ERROR FabricInfo::SetExternallyOwnedOperationalKeypair(P256Keypair * keyPair)
```
Comment (bzbarsky-apple) [inline src/credentials/FabricTable.cpp:344]: As a follow-up, maybe some of the keypair methods should become const, so we can take a `const P256Keypair *` here?

File: src/credentials/FabricTable.h [hunk 5]
```diff
@@ -123,7 +123,30 @@ class DLL_EXPORT FabricInfo
         }
         return mOperationalKey;
     }
-    CHIP_ERROR SetOperationalKeypair(const Crypto::P256Keypair * keyPair);
+
+    /**
+     * Sets the P256Keypair used for this fabric, transferring ownership of the
+     * key to this object by making a copy via the P256Keypair::Serialize and
+     * P256Keypair::Deserialize methods.
```
Comment (bzbarsky-apple) [inline src/credentials/FabricTable.h:130]: This is not transferring ownership.  How about:
```suggestion
     * Sets the P256Keypair used for this fabric.  This will make a copy of the keypair
     * via the P256Keypair::Serialize and P256Keypair::Deserialize methods.
     *
     * The keyPair argument is safe to deallocate once this method returns.
```

File: src/credentials/FabricTable.h [hunk 6]
```diff
@@ -186,7 +209,7 @@ class DLL_EXPORT FabricInfo
         mVendorId       = VendorId::NotSpecified;
         mFabricLabel[0] = '\0';
 
-        if (mOperationalKey != nullptr)
+        if (mHasExternallyOwnedOperationalKey && mOperationalKey != nullptr)
         {
             chip::Platform::Delete(mOperationalKey);
             mOperationalKey = nullptr;
```
Comment (bzbarsky-apple) [inline src/credentials/FabricTable.h:215]:
```suggestion
            chip::Platform::Delete(mOperationalKey);
        }
        mOperationalKey = nullptr;
```

File: src/credentials/FabricTable.cpp [hunk 7]
```diff
@@ -117,7 +117,10 @@ CHIP_ERROR FabricInfo::CommitToStorage(PersistentStorageDelegate * storage)
 
         ReturnErrorOnFailure(writer.Put(kOpKeyVersionTag, kOpKeyVersion));
 
-        ReturnErrorOnFailure(writer.Put(kOpKeyDataTag, ByteSpan(serializedOpKey.Bytes(), serializedOpKey.Length())));
+        if (!mHasExternallyOwnedOperationalKey)
+        {
+            ReturnErrorOnFailure(writer.Put(kOpKeyDataTag, ByteSpan(serializedOpKey.Bytes(), serializedOpKey.Length())));
```
Comment (bzbarsky-apple) [inline src/credentials/FabricTable.cpp:122]: But now `FabricInfo::LoadFromStorage` will fail for this fabric index.  That function needs changes to deal with a missing `kOpKeyDataTag`, no?

File: src/credentials/FabricTable.h [hunk 8]
```diff
@@ -123,7 +123,30 @@ class DLL_EXPORT FabricInfo
         }
         return mOperationalKey;
     }
-    CHIP_ERROR SetOperationalKeypair(const Crypto::P256Keypair * keyPair);
+
```
Comment (mrjerryjohns) [inline src/credentials/FabricTable.h:126]: `GetOperationalKey` creates an operational key if one isn't present, but it should likely consult with `mHasExternallyOwnedOperationalKey` to make that determination. There is a possibility that can be 'true' BUT you have yet to actually fill in the externally provided key (i.e the time window between fabric infos being loaded from storage, to when the Controller is initialized).

See comment above.

File: src/credentials/FabricTable.cpp (hunk 7 above)
Comment (mrjerryjohns) [reply src/credentials/FabricTable.cpp:122]: I'd think that we should:

1. Assume the FabricInfo is externally owned if `kOpKeyDataTag` is not encountered in `LoadFromStorage`.
2. There is now a period of time between `LoadFromStorage` till when `InitControllerNOCChain` is called on the controller instance associated with the respective `FabricInfo` that the fabric info instance is actually _invalid_. I think that should be caught in `GetOperationalKey`, which is the getter that most external consumers use to get at the key. Presumably, that should just return `mOperationalKey` regardless of whether it is null or not if the flag is set to true?

Comment (tcarmelveilleux) [reply src/credentials/FabricTable.cpp:122]: No, because it will never get called with anything in it from a controller that externally manages everything

Comment (mrjerryjohns) [reply src/credentials/FabricTable.cpp:122]: It will get called in all cases I believe, due to this call in `DeviceControllerFactory::InitSystemState` --     

```
ReturnErrorOnFailure(stateParams.fabricTable->Init(params.fabricIndependentStorage));
```

File: src/controller/CHIPDeviceController.h (hunk 1 above)
Comment (tcarmelveilleux) [reply src/controller/CHIPDeviceController.h:255]: Will add a comment, it's because the `AndroidDeviceControllerWrapper` API doesn't allow yet full injection, so we have to do it after the fact...

File: src/credentials/FabricTable.cpp (hunk 2 above)
Comment (tcarmelveilleux) [reply src/credentials/FabricTable.cpp:98]: Will do

File: src/credentials/FabricTable.cpp (hunk 7 above)
Comment (tcarmelveilleux) [reply src/credentials/FabricTable.cpp:122]: @mrjerryjohns It will get called, but you could have it empty at the start everytime, in which case nothing gets loaded.

Comment (bzbarsky-apple) [reply src/credentials/FabricTable.cpp:122]: Discussed this with @tcarmelveilleux.  What we should do is this: when loading a fabric table entry from storage, if it has an externally managed key (which we can either detect via "we did not have a tag for the key in this TLV" or we could store the boolean for "has externally managed key" explicitly") we should just set `mOperationalKey` to nullptr.  Then we modify `FabricInfo::GetOperationalKey` to just return `mOperationalKey` if `mHasExternallyOwnedOperationalKey` without trying to heap-allocate it there.

The callers of `GetOperationalKey()` all do null-checks already, so will correctly handle it being null (typically by doing things like erroring out of CASE session establishment, which is what we want anyway in that situation).

**Review by msandstedt — APPROVED** [review]

File: src/credentials/FabricTable.cpp [hunk 9]
```diff
@@ -213,27 +206,36 @@ CHIP_ERROR FabricInfo::LoadFromStorage(PersistentStorageDelegate * storage)
...
 #endif
+            }
+            VerifyOrReturnError(mOperationalKey != nullptr, CHIP_ERROR_NO_MEMORY);
+            ReturnErrorOnFailure(mOperationalKey->Deserialize(serializedOpKey));
+        }
+        else
+        {
+            // Key was absent: set mOperationalKey to null, for another caller to set
+            // it. This may happen if externally owned.
+            mOperationalKey = nullptr;
```
Comment (bzbarsky-apple) [inline src/credentials/FabricTable.cpp:237]: Do we want to set `mHasExternallyOwnedOperationalKey` here?  See also question about `GetOperationalKey`.

**Review by bzbarsky-apple — COMMENTED** [review]
Still need to fix `GetOperationalKey` to not allocate `mOperationalKey` if it's null and `mHasExternallyOwnedOperationalKey`, right?  Or are we OK with that just allocating a bogus key that will fail further down the line (which I think I am)?

File: src/credentials/FabricTable.cpp (hunk 9 above)
Comment (tcarmelveilleux) [reply src/credentials/FabricTable.cpp:237]: No, we don't want to set it. We want things to be as close as before the change.

**Review by mrjerryjohns — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

File: src/credentials/FabricTable.cpp (hunk 4 above)
Comment (tcarmelveilleux) [reply src/credentials/FabricTable.cpp:343]: Sounds good
